fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/LuceneIndexMaintainer.java

```diff
@@ -0,0 +1,244 @@
+        super(state);
+        this.analyzer = analyzer;
+        this.fieldNames = new ArrayList<>(2);
+        this.state.index.getRootExpression()
+                .normalizeKeyForPositions().forEach( (fke) -> this.fieldNames.add( ((FieldKeyExpression) fke).getFieldName()));
+        this.executor = executor;
+    }
+
+    /**
+     * The scan uses the low element in the range to execute the
+     * MultiFieldQueryParser.
+     *
+     * @param scanType the {@link IndexScanType type} of scan to perform
+     * @param range the range to scan
+     * @param continuation any continuation from a previous scan invocation
+     * @param scanProperties skip, limit and other properties of the scan
+     * @return RecordCursor
+     */
+    @Nonnull
+    @Override
+    public RecordCursor<IndexEntry> scan(@Nonnull IndexScanType scanType, @Nonnull TupleRange range,
+                                         @Nullable byte[] continuation,
+                                         @Nonnull ScanProperties scanProperties) {
+        LOG.trace("scan scanType={}", scanType);
+        Verify.verify(range.getLow() != null);
+        try {
+            final MultiFieldQueryParser parser = new MultiFieldQueryParser(fieldNames.toArray(new String[0]) , analyzer);
+            Query query = parser.parse(range.getLow().getString(0));
+            return new LuceneRecordCursor(executor, scanProperties, state, query, continuation, fieldNames);
+        } catch (Exception ioe) {
+            throw new RuntimeException(ioe);
+        }
+    }
+
+    @Nonnull
+    @Override
+    public <M extends Message> CompletableFuture<Void> update(@Nullable FDBIndexableRecord<M> oldRecord,
+                                                              @Nullable FDBIndexableRecord<M> newRecord) {
+        LOG.trace("update oldRecord={}, newRecord{}", oldRecord, newRecord);
+        return super.update(oldRecord, newRecord);
+    }
+
+    @Nonnull
+    @Override
+    protected <M extends Message> CompletableFuture<Void> updateIndexKeys(@Nonnull final FDBIndexableRecord<M> savedRecord,
+                                                                          final boolean remove,
+                                                                          @Nonnull final List<IndexEntry> indexEntries) {
+        LOG.trace("updateIndexKeys savedRecord={}, remove=={}, indexEntries={}", savedRecord, remove, indexEntries);
+        if (indexEntries.isEmpty()) {
+            return AsyncUtil.DONE;
+        }
+        try {
+            final IndexWriter writer = getOrCreateIndexWriter(state, analyzer);
+            indexEntries.forEach( (entry ) -> {
+                try {
+                    if (remove) {
+                        Query query = SortedDocValuesField.newSlowExactQuery(PRIMARY_KEY_SEARCH_NAME, new BytesRef(savedRecord.getPrimaryKey().pack()));
+                        writer.deleteDocuments(query);
```

Comment:
Are these potentially blocking calls, too? Or do they just queue work and I/O happens at commit time? Mainly for the record; we don't need to fix absolutely everything all at once.

Reply:
This seems to queue work as the change isn't visible until commit is called. Though according to some javadocs deeper in Lucene if the deletes are using too much heap they'll be written to disk. So... both?

Reply:
But mostly happens without I/O, so maybe okay for now.